File: Algorithms/p116_Populating_Next_Right_Pointers_in_Each_Node/p116_Populating_Next_Right_Pointers_in_Each_Node_1.cpp
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
public:
    void connect(TreeLinkNode *root) {
        if(!root || !root->left&&!root->right) return;
        connect(root->left);
        connect(root->right);
        TreeLinkNode* p = root->left;
        TreeLinkNode* q = root->right;
        for(;p&&q;p=p->right,q=q->left){
            p->next=q;
        }
    }
};

File: Algorithms/p116_Populating_Next_Right_Pointers_in_Each_Node/p116_Populating_Next_Right_Pointers_in_Each_Node_1.py
# Definition for binary tree with next pointer.
# class TreeLinkNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution:
    # @param root, a tree link node
    # @return nothing
    def connect(self, root):
        if not root or (not root.left and not root.right):
            return
        self.connect(root.left)
        self.connect(root.right)
        p = root.left
        q = root.right
        while p and q:
            p.next = q
            p = p.right
            q = q.left
